### PR TITLE
Implement commit format

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: repository/src/format.rs [hunk 1]
```diff
@@ -1,31 +1,219 @@
...
+        Commit::ExtraAgendaTransaction(_) => unimplemented!(),
+        Commit::ChatLog(_) => unimplemented!(),
     }
 }
 
-pub fn from_semantic_commit(_semantic_commit: SemanticCommit) -> Result<Commit, String> {
-    todo!()
+/// Converts a semantic commit to a commit.
+///
+/// TODO: retrieve author and timestamp from the commit metadata.,
```
Comment (TomTaehoonKim) [inline repository/src/format.rs:47]:
```suggestion
/// TODO: retrieve author and timestamp from the commit metadata.
```

File: repository/src/format.rs [hunk 2]
```diff
@@ -1,31 +1,219 @@
...
+                        height
+                    ));
+                }
+                Ok(Commit::Block(block_header))
+            }
+            "agenda-proof" => {
+                let agenda_proof: AgendaProof = serde_json::from_str(&semantic_commit.body)?;
+                if height != agenda_proof.height {
+                    return Err(anyhow!(
+                        "agenda proof height mismatch: expected {}, got {}",
```
Comment (TomTaehoonKim) [inline repository/src/format.rs:92]:
```suggestion
                        "agenda-proof height mismatch: expected {}, got {}",
```

File: repository/src/format.rs [hunk 3]
```diff
@@ -1,31 +1,219 @@
...
+                        agenda_proof.height,
+                        height
+                    ));
+                }
+                Ok(Commit::AgendaProof(agenda_proof))
+            }
+            _ => Err(anyhow!("Unknown commit type: {}", commit_type)),
```
Comment (TomTaehoonKim) [inline repository/src/format.rs:99]:
```suggestion
            _ => Err(anyhow!("unknown commit type: {}", commit_type)),
```

File: repository/src/format.rs [hunk 4]
```diff
@@ -1,31 +1,219 @@
...
+        }
+    } else {
+        Ok(Commit::Transaction(Transaction {
+            author: PublicKey::zero(),
+            timestamp: 0,
```
Comment (TomTaehoonKim) [inline repository/src/format.rs:104]: future work? or is it okay this way?

File: repository/src/format.rs [hunk 5]
```diff
@@ -1,31 +1,219 @@
...
+) -> Result<LastFinalizationProof, Error> {
+    let pattern = Regex::new(r"^>fp: (\d+)$").unwrap();
+    let captures = pattern.captures(&semantic_commit.title);
+    if let Some(captures) = captures {
+        let height = captures.get(1).map_or("", |m| m.as_str());
+        let height = height.parse::<u64>()?;
+        let proof: FinalizationProof = serde_json::from_str(&semantic_commit.body)?;
+        Ok(LastFinalizationProof { height, proof })
+    } else {
+        Err(anyhow!("Unknown commit type: {}", semantic_commit.title))
```
Comment (TomTaehoonKim) [inline repository/src/format.rs:133]:
```suggestion
        Err(anyhow!("unknown commit type: {}", semantic_commit.title))
```

File: repository/src/format.rs [hunk 6]
```diff
@@ -1,31 +1,219 @@
...
 pub fn fp_from_semantic_commit(
-    _semantic_commit: SemanticCommit,
-) -> Result<LastFinalizationProof, String> {
-    todo!()
+    semantic_commit: SemanticCommit,
+) -> Result<LastFinalizationProof, Error> {
+    let pattern = Regex::new(r"^>fp: (\d+)$").unwrap();
+    let captures = pattern.captures(&semantic_commit.title);
+    if let Some(captures) = captures {
+        let height = captures.get(1).map_or("", |m| m.as_str());
```
Comment (TomTaehoonKim) [inline repository/src/format.rs:128]:
```suggestion
    let pattern = Regex::new(r"^>(fp): (\d+)$").unwrap();
    let captures = pattern.captures(&semantic_commit.title);
    if let Some(captures) = captures {
        let commit_type = captures.get(1).map(|m| m.as_str()).ok_or_else(|| {
                anyhow!(
                    "Failed to parse commit type from commit title: {}",
                    semantic_commit.title
                )
            })?;
        let height = captures.get(2).map(|m| m.as_str()).ok_or_else(|| {
            anyhow!(
                "Failed to parse commit height from commit title: {}",
                semantic_commit.title
            )
        })?;
```

File: repository/src/format.rs [hunk 7]
```diff
@@ -1,31 +1,219 @@
...
+            author: PublicKey::zero(),
+            timestamp: 0,
+            head: "abc".to_string(),
+            body: "def".to_string(),
+            diff: Diff::None,
+        });
+        assert_eq!(
+            agenda,
+            from_semantic_commit(to_semantic_commit(&agenda)).unwrap()
+        );
```
Comment (TomTaehoonKim) [inline repository/src/format.rs:153]:
```suggestion
        let transaction = Commit::Transaction(Transaction {
            author: PublicKey::zero(),
            timestamp: 0,
            head: "abc".to_string(),
            body: "def".to_string(),
            diff: Diff::None,
        });
        assert_eq!(
            transaction,
            from_semantic_commit(to_semantic_commit(&transaction)).unwrap()
        );
```

File: repository/src/format.rs [hunk 8]
```diff
@@ -1,31 +1,219 @@
...
+            timestamp: 0,
+            commit_merkle_root: Hash256::hash("hello2"),
+            repository_merkle_root: Hash256::hash("hello3"),
+            validator_set: vec![(PublicKey::zero(), 1)],
+            version: "0.1.2".to_string(),
+        });
+        assert_eq!(
+            block_header,
+            from_semantic_commit(to_semantic_commit(&block_header)).unwrap()
+        );
```
Comment (TomTaehoonKim) [inline repository/src/format.rs:189]:
```suggestion
        let block = Commit::Block(BlockHeader {
            height: 3,
            author: PublicKey::zero(),
            prev_block_finalization_proof: vec![TypedSignature::new(
                Signature::zero(),
                PublicKey::zero(),
            )],
            previous_hash: Hash256::hash("hello1"),
            timestamp: 0,
            commit_merkle_root: Hash256::hash("hello2"),
            repository_merkle_root: Hash256::hash("hello3"),
            validator_set: vec![(PublicKey::zero(), 1)],
            version: "0.1.2".to_string(),
        });
        assert_eq!(
            block,
            from_semantic_commit(to_semantic_commit(&block)).unwrap()
        );
```

File: repository/src/format.rs (hunk 4 above)
Comment (junha1) [reply repository/src/format.rs:104]: see line 47

File: repository/src/format.rs (hunk 6 above)
Comment (junha1) [reply repository/src/format.rs]: fixed.

File: repository/src/format.rs [hunk 9]
```diff
@@ -1,31 +1,219 @@
...
+            let title = format!(">block: {}", block_header.height);
+            let body = serde_json::to_string(block_header).unwrap();
+            SemanticCommit {
+                title,
+                body,
+                diff: Diff::None,
+            }
+        }
+        Commit::Transaction(transaction) => SemanticCommit {
+            title: transaction.head.clone(),
```
[the remark this answers is not in the document]
Comment (TomTaehoonKim) [reply repository/src/format.rs:27]: A lot of renaming required so will be handled in future PR

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Comment by YoonhoKim0527** [comment]
LGTM too.

**Review by YoonhoKim0527 — COMMENTED** [review]
LGTM too.

**Pull request merged**